tracer/src/Datadog.Trace/PDBs/SourceLink/AzureDevOpsSourceLinkUrlParser.cs

```diff
@@ -0,0 +1,76 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Collections.Specialized;
+using System.Diagnostics.CodeAnalysis;
+
+#pragma warning disable CS1570
+
+namespace Datadog.Trace.Pdb.SourceLink;
+
+internal class AzureDevOpsSourceLinkUrlParser : SourceLinkUrlParser
+{
+    /// <summary>
+    /// Extract the git commit sha and repository url from a Azure DevOps SourceLink mapping string.
+    /// For example, for the following SourceLink mapping string:
+    ///     https://test.visualstudio.com/test-org/_apis/git/repositories/my-repo/items?api-version=1.0&amp;versionType=commit&amp;version=dd35903c688a74b62d1c6a9e4f41371c65704db8&amp;path=/*
+    /// It will return:
+    ///     - commit sha: dd35903c688a74b62d1c6a9e4f41371c65704db8
+    ///     - repository URL: https://test.visualstudio.com/test-org/_git/my-repo
+    /// </summary>
+    internal override bool TryParseSourceLinkUrl(Uri uri, [NotNullWhen(true)] out string? commitSha, [NotNullWhen(true)] out string? repositoryUrl)
+    {
+        commitSha = null;
+        repositoryUrl = null;
+
+        // Check if the given URI is a valid AzureDevOps SourceLink mapping string.
+        if (!uri.AbsolutePath.Contains("_apis/git/repositories/") ||
+            !uri.Query.Contains("versionType=commit") ||
+            !uri.Query.Contains("version=") ||
+            !uri.Query.Contains("path=/*"))
+        {
+            return false;
+        }
+
+        // Extract the query string of the URI and check if the required parameters exist.
+        var query = ParseQueryString(uri.Query);
+        commitSha = query["version"];
+        if (!IsValidCommitSha(commitSha))
+        {
+            return false;
+        }
+
+        var segments = uri.AbsolutePath.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+        if (segments.Length < 5)
+        {
+            return false;
+        }
+
+        repositoryUrl = $"https://{uri.Host}/{segments[0]}/_git/{segments[4]}";
+
+        return true;
+    }
+
+    private static NameValueCollection ParseQueryString(string queryString)
+    {
+        // We can't use HttpUtility.ParseQueryString because it would mean taking a dependency on System.Web.
+        // Instead, we parse the query string manually by simply splitting on the '&' character,
+        // as in our particular use case, there is no need to decode the values.
+        var query = new NameValueCollection();
+        var pairs = queryString.Split(new[] { '&' }, StringSplitOptions.RemoveEmptyEntries);
+        foreach (var pair in pairs)
+        {
+            var parts = pair.Split(new[] { '=' }, 2);
+            if (parts.Length == 2)
+            {
+                query.Add(parts[0], parts[1]);
+            }
+        }
+
+        return query;
+    }
```

Comment:
Arguably, you're only looking for a single query value, so if you wish you _could_ skip the `NameValueCollection` and all the `string.Split()` entirely.

Can't decide if I hate it 😅 


```suggestion
    private static string GetVersion(string queryString)
    {
        // We can't use HttpUtility.ParseQueryString because it would mean taking a dependency on System.Web.
        // Instead, we parse the query string manually by simply splitting on the '&' character,
        // as in our particular use case, there is no need to decode the values.
        var index = 0;
        while (index > -1)
        {
            var equalsIndex = queryString.IndexOf('=', startIndex: index);

            if (equalsIndex == index + 7 && IsVersion(queryString, index))
            {
                // found version, extract value
                var nextAmpersand = queryString.IndexOf('&', startIndex: equalsIndex);
                var length = nextAmpersand - equalsIndex - 1;
                return nextAmpersand > equalsIndex + 1
                           ? queryString.Substring(equalsIndex + 1, length)
                           : queryString.Substring(equalsIndex + 1);
            }

            index = queryString.IndexOf('&', index) + 1;
        }

        // not found
        return null;

        static bool IsVersion(string queryString, int keyStartIndex)
            => queryString[keyStartIndex] is 'v' or 'V'
            && queryString[keyStartIndex + 1] is 'e' or 'E'
            && queryString[keyStartIndex + 2] is 'r' or 'R'
            && queryString[keyStartIndex + 3] is 's' or 'S'
            && queryString[keyStartIndex + 4] is 'i' or 'I'
            && queryString[keyStartIndex + 5] is 'o' or 'O'
            && queryString[keyStartIndex + 6] is 'n' or 'N';
    }
```

Reply:
Your proposal is definitely a **lot** better than mine performance-wise and allocation-wise, but I tend to think the original code was more readable. Since I expect this method to be executed _once_ in the application's lifetime, and find it hard to imagine future scenarios where it'll be executed much more than that, I think readability should win. WDYT?

Reply:
I'm fine with it either way 😄 My initial thought was "this is probably simple to do without `Split`", and I went down the rabbit hole... "simple" is hard apparently 🤣 

Mine would require tests, yours doesn't. I think that says enough from my point of view, leave it as-is 😉